src/Features/Core/Portable/RemoveUnusedMembers/AvoidUnusedMembersDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,199 @@
+                    {
+                        // Check if the underlying member is neither read nor a readable reference to the member is taken.
+                        if (symbolValueUsageStateMap.TryRemove(member, out var valueUsageInfo) &&
+                            (valueUsageInfo & (ValueUsageInfo.Read | ValueUsageInfo.ReadableRef)) == 0)
+                        {
+                            Debug.Assert(IsCandidateSymbol(member));
+
+                            var option = GetCodeStyleOption(member, symbolEndContext);
+                            if (option != null && option.Value)
+                            {
+                                // Give appropriate message based on usage:
+                                //  1. No read or writes:   Type '{0}' has an unused private member '{1}' which can be removed.
+                                //  2. Only writes:         Type '{0}' has a private member '{1}' which can be removed as the value assigned to it is never used.
+
+                                var rule = (valueUsageInfo & (ValueUsageInfo.Write | ValueUsageInfo.WritableRef)) != 0 ?
+                                    _noReadsUnnecessaryWithSuggestionDescriptor :
+                                    UnnecessaryWithSuggestionDescriptor;
+
+                                var additionalLocations = member.Locations.Skip(1).ToImmutableArrayOrEmpty();
+                                var diagnostic = DiagnosticHelper.Create(
+                                    rule,
+                                    member.Locations[0],
+                                    option.Notification.Severity,
+                                    additionalLocations,
+                                    properties: ImmutableDictionary.CreateRange(new[] {
+                                        new KeyValuePair<string, string>(UnunsedMemberNameProperty, member.Name),
+                                        new KeyValuePair<string, string>(UnunsedMemberKindProperty, member.Kind.ToString()) }),
+                                    member.ContainingSymbol.Name,
+                                    member.Name);
+                                symbolEndContext.ReportDiagnostic(diagnostic);
+                            }
+                        }
+                    }
+                }
+
+                bool IsCandidateSymbol(ISymbol memberSymbol)
+                {
+                    if (memberSymbol.DeclaredAccessibility == Accessibility.Private &&
+                        !memberSymbol.IsImplicitlyDeclared)
+                    {
+                        // Do not track accessors, as we will track the associated symbol.
+                        switch (memberSymbol.Kind)
+                        {
+                            case SymbolKind.Method:
+                                return ((IMethodSymbol)memberSymbol).AssociatedSymbol == null &&
+                                    memberSymbol != entryPoint;
```

Comment:
the entryPoint method needs to be called 'Main' IIRC (though definitely check on that).  We should avoid this expensive work if it's not something called 'Main'.  Or, alternatively, just assume that any method called 'Main' is used.

Reply:
I simplified the check to compare name, return type and IsStatic.

---
In reply to: [212734091](https://github.com/dotnet/roslyn/pull/29511#discussion_r212734091) [](ancestors = 212734091)